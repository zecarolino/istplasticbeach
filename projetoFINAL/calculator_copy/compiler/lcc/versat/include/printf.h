#ifndef _PRINTF_H_
#define _PRINTF_H_

#include "printi.h"
#include "puts.h"
#include "stdarg.h"

int printf(const char *fmt, ...)
{
  va_list ap;
  va_start(ap, fmt);
  while (*fmt != 0) {
    if (*fmt == '%') {
      fmt++;
      if (*fmt == 'd') printi(va_arg(ap,int), 10);
      else if (*fmt == 's') puts(va_arg(ap,char*));
      else if (*fmt == 'c') putchar(va_arg(ap,int));
      else putchar(*fmt);
    }
    else putchar(*fmt);
    fmt++;
  }
  va_end(ap);
  return 1;
}

#endif /* _PRINTF_H_ */
