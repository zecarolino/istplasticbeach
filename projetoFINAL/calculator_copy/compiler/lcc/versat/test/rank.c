/*
 * Fundamental to any permutation-generation algorithm is a notion of order,
 * the sequence in which the permutations are constructed, from first to last.
 * The most natural generation order is lexicographic, the order they would
 * appear if they were sorted numerically. Lexicographic order for n=3 is
 * {1,2,3}, {1,3,2}, {2,1,3}, {2,3,1}, {3,1,2}, {3,2,1}.
 * rank(p) = (p1 - 1) * (|p| - 1)! + rank({p2, ..., p|p|})
 * gives the position of 'p' in the given generation lexicographic order.
 * Getting this right means relabeling the elements of the smaller permutation
 * to reflect the deleted first element.
 */
int factorial(int n)
{
  if (n < 2) return 1;
  return n * factorial(n-1);
}

int rank(int *num, int siz)
{
  int i;

  if (siz < 1) return 0;

  /* num[0] gets removed so re-number all number above him, i.e,
   * make all remaining numbers in the range 1..(siz-1)
   */
  for (i = 1; i < siz; i++)
    if (num[i] > num[0])
      num[i]--;

  return (num[0] - 1) * factorial(siz - 1) + rank(num+1, siz-1);
}

#include "printf.h"
int main()
{
  int i, sum, n[10], j = 0, cnt;

  n[0] = 3;
  n[1] = 1;
  n[2] = 2;
  cnt = 3;

  for (i = sum = 0; i < cnt; i++, j+=i)
    sum += n[i];

  if (sum == 0 || sum != j) {
    printf("invalid sequence: it must contain all non-repeated numbers from 1 to n\n");
    return 1;
  }

  printf("rank{", sum, j);
  for (i = 0; i < cnt; i++)
    printf(" %d", n[i]);
  printf(" } = %d\n", rank(n, i));

  return 0;
}
