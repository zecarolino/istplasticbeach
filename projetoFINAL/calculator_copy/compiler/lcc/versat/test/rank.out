VCD info: dumpfile xtop.vcd opened for output.
rank{ 3 1 2 } = 4
0
Catched Trap at time  9010 clock cycles:
